src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/BufferBlockCopyLengthAnalyzer.cs

```diff
@@ -0,0 +1,168 @@
+                    ParameterInfo.GetParameterInfo(arrayType, false, 1, false),
+                    ParameterInfo.GetParameterInfo(int32Type, false, 1, false),
+                    ParameterInfo.GetParameterInfo(arrayType, false, 1, false),
+                    ParameterInfo.GetParameterInfo(int32Type, false, 1, false),
+                    ParameterInfo.GetParameterInfo(int32Type, false, 1, false)};
+
+                IMethodSymbol? blockCopyMethod = bufferType
+                   .GetMembers("BlockCopy")
+                   .OfType<IMethodSymbol>()
+                   .GetFirstOrDefaultMemberWithParameterInfos(blockCopyParameterInfo);
+
+                if (blockCopyMethod is null)
+                {
+                    return;
+                }
+
+                ISymbol arrayLengthProperty = arrayType
+                    .GetMembers("Length")
+                    .FirstOrDefault(p => p is IPropertySymbol);
+
+                context.RegisterOperationAction(context =>
+                {
+                    var invocationOperation = (IInvocationOperation)context.Operation;
+
+                    if (!invocationOperation.TargetMethod.Equals(blockCopyMethod, SymbolEqualityComparer.Default))
+                    {
+                        return;
+                    }
+
+                    ImmutableArray<IArgumentOperation> arguments = IOperationExtensions.GetArgumentsInParameterOrder(invocationOperation.Arguments);
+
+                    IArgumentOperation sourceArgument = arguments[0];
+                    IArgumentOperation destinationArgument = arguments[2];
+                    IArgumentOperation countArgument = arguments[4];
+
+                    bool CheckArrayLengthLocalReference(IArgumentOperation targetArgument, IPropertyReferenceOperation lengthPropertyArgument)
+                    {
+                        if (targetArgument.Value is IConversionOperation targetArgValue)
+                        {
+                            if (lengthPropertyArgument.Instance.GetReferencedMemberOrLocalOrParameter() == targetArgValue.Operand.GetReferencedMemberOrLocalOrParameter())
+                            {
+                                IArrayTypeSymbol countArgumentArrayTypeSymbol = (IArrayTypeSymbol)lengthPropertyArgument.Instance.Type;
+                                if (countArgumentArrayTypeSymbol.ElementType.SpecialType != SpecialType.System_Byte &&
+                                countArgumentArrayTypeSymbol.ElementType.SpecialType != SpecialType.System_SByte)
+                                {
+                                    return true;
+                                }
+                            }
+                        }
+                        else if (targetArgument.Value is ILocalReferenceOperation local)
+                        {
+                            if (lengthPropertyArgument.Instance.GetReferencedMemberOrLocalOrParameter() == local.Local)
```

Comment:
Do not need to check the element type of the array?

Reply:
I defer to @buyaa-n and @pgovind here.

Reply:
We should not warn for such cases, not sure if we can report different diagnostic with lower severity like`DiagnosticSeverity.Hidden` for this case (I know we can with different diagnostic ID but not sure if we can with same diagnostic id) CC @mavasani @jmarolf @sharwell

Reply:
Opened #5371 to track this issue. The unit test for this case is being skipped for now.